Optimize WAL storage in safekeeper

petuhovskiy opened the pull request:
When several AppendRequest's can be read from socket without blocking,
they are processed together and fsync() to segment file is only called
once. Segment file is no longer opened for every write request, now
last opened file is cached inside PhysicalStorage. New metric for WAL
flushes was added to the storage, FLUSH_WAL_SECONDS. More errors were
added to storage for non-sequential WAL writes, now write_lsn can be
moved only with calls to truncate_lsn(new_lsn).

New messages have been added to ProposerAcceptorMessage enum. They
can't be deserialized directly and now are used only for optimizing
flushes. Existing protocol wasn't changed and flush will be called for
every AppendRequest, as it was before.

This PR replaces #1266, as a cleaner version of the same optimization. Closes #1144.
I'll post test results here when they're ready.

petuhovskiy commented on walkeeper/src/receive_wal.rs:
```diff
@@ -100,16 +95,81 @@ impl<'pg> ReceiveWalConn<'pg> {
+    }
+
+    fn recv_msg(&mut self) -> Result<ProposerAcceptorMessage> {
+        self.msg_rx.recv().context("failed to receive message")
```
I wanted to replace error here with original read error from thread and tried to use `self.read_thread.join().err().unwrap()` for that, but it didn't work. Does anyone know how to make it work?

arssher replied:
I think because you need to go one level deeper: `join()` returns `Error` only if thread panicked. In our case it would be typically `Ok<Result<()>>` with inner result containing actual error.

petuhovskiy commented on walkeeper/src/wal_storage.rs:
```diff
@@ -339,69 +374,69 @@ impl Storage for PhysicalStorage {
-        if self.write_lsn == Lsn(0) {
-            return Ok(());
-        }
+        // TODO: cross check divergence point?
```
@arssher do you remember what this comment was about? It looks like some check should be added here, but I don't remember what exactly should be checked.

arssher replied:
The idea was to repeat divergence point search like in SendProposerElected to ensure we truncate at right position. The check should be done before calling `truncate_wal`.

petuhovskiy commented:
Here are results of the test from #1144, to compare results with #1266.
Code to test the results of this PR: https://github.com/zenithdb/zenith/tree/test-perf-pr-1318
Same test backported to main: https://github.com/zenithdb/zenith/tree/test-perf-backport-1318

These results are for local run on my machine, which has relatively slow fsync (2364 ops/sec, 423 usecs/op) and not very powerful 8 core CPU. I've added "per fsync" metrics, which are useful for disk usage comparison.

 | Test info  | pgbench -i -s 200 (time) | pgbench -N -T 60 | init wal_bytes, per fsync | bench wal_bytes, per fsync | bench txes, per fsync |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| NO PATCH safekeeper + wp (fsync=on, no checkpoints)  | 108 s | 1177 TPS | 85kb | 436 b | 1 tx / fsync |
| PATCHED safekeeper + wp (fsync=on, no checkpoints)  | 87 s | 2574 TPS | 417kb | 6314 b | 15.4 tx / fsync |
| vanilla repl (fsync=on, no checkpoints) | 146 s | 4215 TPS | 256kb | 5552 b | 14 tx / fsync |

These results look the same as in #1266.

petuhovskiy commented on walkeeper/src/wal_storage.rs:
```diff
@@ -135,128 +159,136 @@ impl PhysicalStorage {
+        if !self.conf.no_sync {
+            self.metrics
+                .flush_wal_seconds
+                .observe_closure_duration(|| file.sync_all())?;
```
sync_data can be used here

petuhovskiy commented:
In a more usual 1wp+3sk EC2 test (#799) results are also similar to previous PR #1266, safekeepers are faster than synchronous replication:

| Code branch | Test info  | pgbench -i -s 400 (time) | pgbench TPS | init wal_bytes, per fsync | bench wal_bytes, per fsync | bench txes, per fsync | link to gist |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| main | vanilla 1primary(fsync=off)+3replica(fsync=on) | 130 s | 14237 TPS | 800 kb | 1948 b | 6.7 tx / fsync | https://gist.github.com/petuhovskiy/35134419ccb0f287046a6b1c9a6150b8 |
| this PR | 1wp(fsync=off)+3sk(fsync=on) | 92 s | 20400 TPS | 209kb | 1309 b | 4.5 tx / fsync | https://gist.github.com/petuhovskiy/c16668bdfea05acf39a8154b81ac756b |

Interesting that fsync calls count is very different between postgres synchronous replicas when pgbench -i is done, that is visible in gist report.

arssher commented on walkeeper/src/receive_wal.rs:
```diff
@@ -100,16 +95,80 @@ impl<'pg> ReceiveWalConn<'pg> {
+                    let msg = ProposerAcceptorMessage::parse(copy_data)?;
+                    msg_tx.send(msg)?;
+                }
+                // msg_rx will be dropped here, this will also close msg_tx
```
rx and tx should be swapped

arssher commented on walkeeper/src/receive_wal.rs:
```diff
@@ -100,16 +95,80 @@ impl<'pg> ReceiveWalConn<'pg> {
-            if let Some(reply) = reply {
-                self.write_msg(&reply)?;
+            if matches!(next_msg, Some(ProposerAcceptorMessage::AppendRequest(_))) {
+                // poll messages without blocking
```
Let's extend the comment to 'write data without flushing while it is readily available'.

arssher commented on walkeeper/src/wal_storage.rs:
```diff
@@ -339,69 +383,67 @@ impl Storage for PhysicalStorage {
+        }
+
+        if let Some(mut unflushed_file) = self.file.take() {
+            self.fdatasync_file(&mut unflushed_file)?;
```
Option can be matched with `ref` keyword to its contents to avoid taking/returning it around.

petuhovskiy replied:
I need to call `self.fdatasync_file` in this if, so with `ref` borrow I get an error:
```
cannot borrow `*self` as immutable because it is also borrowed as mutable
immutable borrow occurs hererustc[E0502](https://doc.rust-lang.org/error-index.html#E0502)
```

arssher commented on walkeeper/src/wal_storage.rs:
```diff
@@ -294,29 +336,31 @@ impl Storage for PhysicalStorage {
 
-        // WAL is written and flushed, updating lsns
-        self.write_lsn = startpos + buf.len() as u64;
+        // WAL is written and flushed, updating LSNs
```
it is not flushed yet

arssher commented on walkeeper/src/wal_storage.rs:
```diff
@@ -135,128 +159,146 @@ impl PhysicalStorage {
-        let mut partial;
-        let mut start_pos = startpos;
-        const ZERO_BLOCK: &[u8] = &[0u8; XLOG_BLCKSZ];
+        if self.write_lsn != pos {
```
How this can be true?

petuhovskiy replied:
Shouldn't be, possible only if someone if using private API directly, functions like `write_exact`.

petuhovskiy commented:
> These results are for local run on my machine, which has relatively slow fsync (2364 ops/sec, 423 usecs/op) and not very powerful 8 core CPU. I've added "per fsync" metrics, which are useful for disk usage comparison.

I've updated PR to use fdatasync in most places, and ran this test again. It seems that my machine has also much slower fsync, compared to fdatasync, and now results are almost the same:

 | Test info  | pgbench -i -s 200 (time) | pgbench -N -T 60 | init wal_bytes, per fsync | bench wal_bytes, per fsync | bench txes, per fsync |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| PATCHED safekeeper + wp (fsync=on, no checkpoints), using fdatasync  | 99 s | 4041 TPS | 301kb | 5439 b | 13.8 tx / fsync |
| vanilla repl (fsync=on, no checkpoints) | 149 s | 4170 TPS | 257kb | 5674 b | 14.5 tx / fsync |
